### PR TITLE
Pandaskill

### PR DESCRIPTION
**Pull request by dotXem**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,115 @@
...
+├── expert_surveys/ # files used for the expert evaluation of the models
+├── preprocessing/ # features extracted from the raw data, used in the visualization app
+└── raw/ # raw data used to produce the experimental results
+```
+<details>
+  <summary>Details for the `raw` subfolder content</summary>
+
+- `game_metadata.csv`: metadata of the games
+    - `game_id`: ID of the game
+    - `date`: date
```
Comment (JordanPeltier) [inline README.md]: precise the format of the date here

File: README.md [hunk 2]
```diff
@@ -1,4 +1,115 @@
...
+    - `match_id`: ID of the match (e.g., a BO5 is a match of max 5 games)
+    - `tournament_id`: ID of the tournament
+    - `tournament_name`: name of the tournament (e.g., Playoffs)
+    - `serie_id`: ID of the serie
+    - `serie_name`: name of the serie (e.g., LCK Summer 2024)
+    - `league_id`: ID of the league
+    - `league_name`: name of the league (e.g., LCK)
+
+Note: every game can be included in a tree structure such that: `League > Serie > Tournament > Match > Game`.
```
Comment (JordanPeltier) [inline README.md]: I would reverse and use `isin` instead of `>`

File: README.md [hunk 3]
```diff
@@ -1,4 +1,115 @@
...
+    - `team_id`: ID of the player's team
+    - `team_name`: name of the player's team
+    - `team_acronym`: acronym of the player's team
+    - `role`: role of the player (e.g., Mid)
+    - `win`: whether the player has won the game or not
+    - `game_length`: length of the game in seconds
+    - `champion_name`: name of the Champion played by the player
+    - `team_kills`: total number of champion kills of the player's team
+    - `tower_kills`: total number of tower kills of the player's team
+    - `inhibitor_kills`: total number of inhibitor kills of the player's team
```
Comment (JordanPeltier) [inline README.md]: precise that if the same inhibitor is destroyed multiple times, then it is counted multiple times here

File: README.md [hunk 4]
```diff
@@ -1,4 +1,115 @@
...
+# Data
+
+All the data needed to reproduce the results are located in the `pandaskill/artifacts/data/` folder. More specifically:
+```
+pandaskill/artifacts/data/
+├── app/ # performance scores and skill ratings of PandaSkill, used in the visuzalization app
+├── expert_surveys/ # files used for the expert evaluation of the models
+├── preprocessing/ # features extracted from the raw data, used in the visualization app
+└── raw/ # raw data used to produce the experimental results
+```
```
Comment (JordanPeltier) [inline README.md:48]: I think I would put this detailed description of the dataset in the `artifacts`
folder and link it here. That would make the root README more digest

File: README.md [hunk 5]
```diff
@@ -1,4 +1,115 @@
-# README
+# Description
```
Comment (JordanPeltier) [inline README.md:9]: Maybe add a table of content at the beginning ?

File: README.md [hunk 6]
```diff
@@ -1,4 +1,115 @@
...
+<details>
+  <summary>Details for the `raw` subfolder content</summary>
+
+- `game_metadata.csv`: metadata of the games
+    - `game_id`: ID of the game
+    - `date`: date
+    - `match_id`: ID of the match (e.g., a BO5 is a match of max 5 games)
+    - `tournament_id`: ID of the tournament
+    - `tournament_name`: name of the tournament (e.g., Playoffs)
+    - `serie_id`: ID of the serie
```
Comment (JordanPeltier) [inline README.md]: `serie` as the singular of series does not exist. Is it painful to replace `serie` by `series` everywhere ?

File: pandaskill/app/streamlit_app.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+from pandaskill.app.leaderboard_page import display_leaderboard_page
+from pandaskill.app.data import get_all_data
+from pandaskill.app.region_page import display_region_page
+from pandaskill.app.player_team_page import display_player_team_page
+from pandaskill.app.game_page import display_game_page
+
+st.set_page_config(layout="wide", page_icon=":panda_face:", page_title="PandaSkill")
+st.title("PandaSkill")
+st.info(
+    "PandaSkill is an app showing pro League of Legend player performances and skill ratings, following the methodology described in the paper [PandaSkill - Player Performance and Skill Rating in Esports: Application to League of Legends](https://arxiv.org/abs/2109.15098) by [PandaScore](https://pandascore.co/)."
```
Comment (JordanPeltier) [inline pandaskill/app/streamlit_app.py:11]: Shall we add a link to the licence here ?

File: pandaskill/experiments/general/utils.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        data = pd.concat([data, performance_scores_df], axis=1)
+
+    if skill_rating_path:
+        skill_rating_df = pd.read_csv(skill_rating_path, index_col=(0,1))
+        data = pd.concat([data, skill_rating_df], axis=1)
+
+    data = data.sort_values(by=["date", "win"], ascending=[True, False])
+    
+    if drop_na:
+        data = data.dropna()
```
Comment (JordanPeltier) [inline pandaskill/experiments/general/utils.py:52]: What is the impact of removing NAs on the dataset ? 
Does it change some player ratings or something like this ?

File: pandaskill/app/leaderboard_page.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
+import streamlit as st
+import datetime as dt
+import numpy as np
+from pandaskill.experiments.skill_rating.ranking import create_global_player_ranking
+from pandaskill.experiments.general.utils import ALL_REGIONS
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+def display_leaderboard_page(data):    
+    st.header("Leaderboard")
```
Comment (JordanPeltier) [inline pandaskill/app/leaderboard_page.py]: I would add a docstring for each of the tab, explaining the purpose of it

File: pandaskill/app/player_team_page.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+            team_data = data.loc[data["team_id"].isin(selected_team_ids)].copy()
+            team_data["entity_name"] = team_data["team_name"]
+            ratings = team_data.groupby(["entity_name", "game_id", "date"]).agg(
+                serie_name=("serie_name", "first"),
+                tournament_name=("tournament_name", "first"),
+                pscore=("pscore", "mean"),
+                skill_rating_mu=("skill_rating_mu", "mean"),
+                skill_rating_sigma=("skill_rating_sigma", lambda x: np.sqrt(np.mean(np.square(x)))),
+            ).reset_index()
+            ratings["skill_rating"] = ratings["skill_rating_mu"] - 3 * ratings["skill_rating_sigma"]
```
Comment (JordanPeltier) [inline pandaskill/app/player_team_page.py]: link to the paper justifying why using this equation would be good here

Comment (JordanPeltier) [inline pandaskill/app/player_team_page.py]: also I feel like it's missing line breaks in this function. For instance this line, computing actual skill rating from mu and sigma could be highlighted with a line break (in addition to a comment)

File: pandaskill/app/player_team_page.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
+import streamlit as st
+import datetime as dt
+import pandas as pd
+import altair as alt
+import numpy as np
+
+def display_player_team_page(data):
```
Comment (JordanPeltier) [inline pandaskill/app/player_team_page.py:8]: 75 lines seems a bit too much for only one function

File: pandaskill/app/leaderboard_page.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    with setting_columns[2]:
+        all_regions_choice = ["All"] + ALL_REGIONS
+        region = st.selectbox("Region", all_regions_choice, 0)
+    with setting_columns[3]:    
+        all_roles_choice = ["All"] + data["role"].unique().tolist()
+        role = st.selectbox("Role", all_roles_choice, 0)
+
+    since = date - dt.timedelta(days=30*6)
+    since = since.strftime("%Y-%m-%d")
+    min_nb_games = 10
```
Comment (JordanPeltier) [inline pandaskill/app/leaderboard_page.py]: Why not putting this constant in a utils file and in capital letters ?

File: pandaskill/app/leaderboard_page.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        "player_name": "Player",
+        "team_name": "Team",
+        "role": "Role",
+        "region": "Region",
+        "nb_games": "Nb Games",
+        "last_game_date": "Last Game Date",
+        "pscore": "PScore",
+        "skill_rating_mu": "Skill Rating Mu",
+        "skill_rating_sigma": "Skill Rating Sigma",
+        "skill_rating": "Skill Rating (99.7% CI)"
```
Comment (JordanPeltier) [inline pandaskill/app/leaderboard_page.py:52]: the 99.7% not explained here could be weird for someone not familiar with stats. I would add a link to https://fr.wikipedia.org/wiki/R%C3%A8gle_68-95-99,7

File: pandaskill/app/leaderboard_page.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    setting_columns = st.columns(4)
+    with setting_columns[0]:
+        ranking_type = st.selectbox("Ranking type", ["Player", "Team"], 0)
+    with setting_columns[1]:
+        date = st.date_input("Date", date_default, key="leaderboard_date")
+    with setting_columns[2]:
+        all_regions_choice = ["All"] + ALL_REGIONS
+        region = st.selectbox("Region", all_regions_choice, 0)
+    with setting_columns[3]:    
+        all_roles_choice = ["All"] + data["role"].unique().tolist()
```
Comment (JordanPeltier) [inline pandaskill/app/leaderboard_page.py]: is your app fast ? (I did not run it locally)
if not you could apply all filter in one short instead of doing it sequentially

File: pandaskill/app/leaderboard_page.py (hunk 12 above)
Comment (JordanPeltier) [reply pandaskill/app/leaderboard_page.py]: Also are you sure it will not be relevant for the end user to tune this parameter manually ?

File: pandaskill/experiments/general/metrics.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
+import numpy as np
+import pandas as pd
+from sklearn.metrics import brier_score_loss
+
+def compute_brier_score(y_true: np.ndarray, y_prob: np.ndarray) -> float:
```
Comment (JordanPeltier) [inline pandaskill/experiments/general/metrics.py]: This function is not usefull imo

File: README.md (hunk 4 above)
Comment (dotXem) [reply README.md:48]: the detailed description of the data is within a collapsible section, so I think it would be fine.

It looks like this

image

File: pandaskill/app/leaderboard_page.py (hunk 14 above)
Comment (dotXem) [reply pandaskill/app/leaderboard_page.py]: app is fast

File: pandaskill/experiments/general/utils.py (hunk 8 above)
Comment (dotXem) [reply pandaskill/experiments/general/utils.py:52]: a total of 71 games are dropped for various reasons (missing stats, missing player), so yeah it impacts the rating, but very slightly.
